File: Projects/Sockets/Environment/mines.c
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "mines.h"

typedef struct 
{
  int START;			/* Start marker in grid */
  int GOAL;			/* End marker in grid */
  int LAND;			/* Free space in grid */
  int OBSTACLE;	                /* Obstical in grid */
  int MINE;                     /* Mine in grid */
  int row;			/* Number of rows in grid */
  int col;			/* Number of columns in grid */
  int startRow;	         	/* Starting position */
  int startCol;   		/* Starting position */
  int agentRow;		        /* Agents current position */
  int agentColumn;	        /* Agents current position */
} mine_env;

Observation o;
mine_env M;
Reward_observation ro;
int env_terminal;

int env_map[6][19] = 
{ 
  { 3, 3, 3, 3, 3, 3 ,3 ,3, 3, 3 ,3 ,3 ,3, 3, 3, 3, 3, 3 }, 
  { 3, 2, 2, 2, 2, 2, 2, 4, 4, 2, 2, 2, 0, 2, 2, 2, 2, 3 }, 
  { 3, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3 },
  { 3, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 2, 2, 2, 2, 3, 3 },
  { 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 3 },
  { 3, 3, 3, 3, 3, 3 ,3 ,3, 3, 3 ,3 ,3 ,3, 3, 3, 3, 3, 3 }
};

int getPosition()
{
  if (env_map[M.agentRow][M.agentColumn] != M.GOAL && env_map[M.agentRow][M.agentColumn] != M.MINE)
  {    
    /* The episode terminates if the agent hits a mine */
    return M.agentRow*M.col + M.agentColumn;
  }
  else
  {
    env_terminal = 1;
    return -1;
  }
}

void getNextPosition(Action a)
{
  /* When the move would result in hitting an obstacles, the agent simply doesn't move */
  int newRow = M.agentRow;
  int newColumn = M.agentColumn;
  
  if (a.intArray[0] == 0)
    newColumn = M.agentColumn - 1;
  else if (a.intArray[0] == 1)
    newColumn = M.agentColumn + 1;
  else if (a.intArray[0] == 2)
    newRow = M.agentRow - 1;
  else if (a.intArray[0] == 3)
    newRow = M.agentRow + 1;
  
  if(newRow >= M.row || newRow < 0 || newColumn >= M.col || newColumn < 0)
  {
    M.agentColumn = M.agentColumn;
    M.agentRow = M.agentRow;
  }
  else if (env_map[newRow][newColumn] != M.OBSTACLE)
  {
    M.agentRow = newRow;
    M.agentColumn = newColumn;
  }
}

Reward getReward()
{
  if (env_map[M.agentRow][M.agentColumn] == M.GOAL){
    return 10;}
  else if (env_map[M.agentRow][M.agentColumn] == M.MINE){
    return -10;}
  else
    return -1;
}

/* RL-Glue Interface */

Task_specification env_init()
{    
  static char Task_spec[100] = {0};
  char temp[50] = {0};

  M.START = 0;
  M.GOAL = 1;
  M.LAND = 2;
  M.OBSTACLE = 3;
  M.MINE = 4;
  
  M.row = 5;
  M.col = 18;
  M.startRow= 1;
  M.startCol = 12;
  M.agentRow = M.startRow;
  M.agentColumn = M.startCol;
  
  o.numInts = 10;
  o.numDoubles = 10;

  o.intArray = (int*)calloc(o.numInts,sizeof(int));
  o.doubleArray = (double*)calloc(o.numDoubles, sizeof(double));

  /* Return task specification */
  
  strcpy(Task_spec,"1:e:1_[i]_[0,");
  sprintf(temp,"%d", M.row*M.col-1);
  strcat(Task_spec,temp);
  sprintf(temp,"]:1_[i]_[0,");
  strcat(Task_spec,temp);
  sprintf(temp,"%d]",4-1);
  strcat(Task_spec,temp); 
  return Task_spec;
}

Observation env_start()
{   
  int i = 0;
  int r = 0, c = 0;

  env_terminal = 0;
  env_map[M.startRow][M.startCol] = M.LAND;
  
  while(env_map[r][c] != M.LAND)
  {
    r =  rand()% M.row;
    c = rand()% M.col;
  }

  M.startRow = r;
  M.startCol = c;
  env_map[M.startRow][M.startCol] = M.START;    
  
  M.agentColumn =  M.startCol;
  M.agentRow = M.startRow;

  for (i = 0; i < 10; ++i)
  {
    o.intArray[i] = -i;
    o.doubleArray[i] = -i - 10;
  }

  o.intArray[0] = M.startRow * M.col + M.startCol;
  
  return o;
}

Reward_observation env_step(Action a)
{    
  int i = 0;

  getNextPosition(a); /* getNextPosition will update the values of agentRow and agentColumn */
 
  o.intArray[0] = getPosition();
  
  ro.o = o;
  ro.r = getReward();
  
  if(env_terminal) /* end of episode? */
    ro.terminal = 1;
  else
    ro.terminal = 0;
  
  return ro;
}

void env_cleanup()
{
  free(o.intArray);
  free(o.doubleArray);
}

void env_set_state(State_key sk)
{
}
     
void env_set_random_seed(Random_seed_key rsk)
{
}

State_key env_get_state()
{
  State_key theKey;
  return theKey;
}

Random_seed_key env_get_random_seed()
{
  Random_seed_key theKey;
  return theKey;
}

File: Projects/Python/Environment/mines.py
import random
import sys
from RL_common import *

class MineEnv:
	START = 0					# Start marker in grid
	GOAL = 0					# End marker in grid
	LAND = 0					# Free space in grid
	OBSTACLE = 0			# Obstical in grid
	MINE = 0					# Mine in grid
	row = 0						# Number of rows in grid
	col = 0						# Number of columns in grid
	startRow = 0			# Starting position
	startCol = 0			# Starting position
	agentRow = 0			# Agent's current position
	agentColumn = 0		# Agent's current position

mine_observation = Observation()
M = MineEnv()
mine_ro = Reward_observation()
mine_terminal = 0

env_map = [ [ 3, 3, 3, 3, 3, 3 ,3 ,3, 3, 3 ,3 ,3 ,3, 3, 3, 3, 3, 3 ], 
						[ 3, 2, 2, 2, 2, 2, 2, 4, 4, 2, 2, 2, 0, 2, 2, 2, 2, 3 ], 
						[ 3, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3 ],
						[ 3, 2, 2, 2, 2, 2, 2, 2, 2, 4, 4, 4, 2, 2, 2, 2, 3, 3 ],
						[ 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 3 ],
						[ 3, 3, 3, 3, 3, 3 ,3 ,3, 3, 3 ,3 ,3 ,3, 3, 3, 3, 3, 3 ]]

def getPosition():
	global mine_terminal
	if env_map[M.agentRow][M.agentColumn] != M.GOAL and env_map[M.agentRow][M.agentColumn] != M.MINE:
		# The episode terminates if the agent hits a mine
		return M.agentRow*M.col + M.agentColumn
	else:
		mine_terminal = 1
		return -1

def getNextPosition(action):
  # When the move would result in hitting an obstacles, the agent simply doesn't move
  newRow = M.agentRow
  newColumn = M.agentColumn
  
  if action.intArray[0] == 0:
    newColumn = M.agentColumn - 1
  elif action.intArray[0] == 1:
    newColumn = M.agentColumn + 1
  elif action.intArray[0] == 2:
    newRow = M.agentRow - 1
  elif action.intArray[0] == 3:
    newRow = M.agentRow + 1
  
  if newRow >= M.row or newRow < 0 or newColumn >= M.col or newColumn < 0:
    M.agentColumn = M.agentColumn
    M.agentRow = M.agentRow
  elif env_map[newRow][newColumn] != M.OBSTACLE:
    M.agentRow = newRow
    M.agentColumn = newColumn

def getReward():
	if env_map[M.agentRow][M.agentColumn] == M.GOAL:
		return 10
	elif env_map[M.agentRow][M.agentColumn] == M.MINE:
		return -10
	else:
		return -1

def env_init():
  M.START = 0
  M.GOAL = 1
  M.LAND = 2
  M.OBSTACLE = 3
  M.MINE = 4
  
  M.row = 5
  M.col = 18
  M.startRow= 1
  M.startCol = 12
  M.agentRow = M.startRow
  M.agentColumn = M.startCol
  
  mine_observation.numInts = 1
  mine_observation.numDoubles = 0

  mine_observation.doubleArray = [];

  random.seed(0)

  # Return task specification
  
  Task_spec = "1:e:1_[i]_[0,%d]:1_[i]_[0,%d]" % (M.row*M.col-1, 4-1)

  return Task_spec;

def env_start():
	r = 0
	c = 0

	mine_terminal = 0
	env_map[M.startRow][M.startCol] = M.LAND

	while env_map[r][c] != M.LAND:
		r = random.randint(0,2) % M.row
		c = random.randint(0,2) % M.col

	M.startRow = r
	M.startCol = c
	env_map[M.startRow][M.startCol] = M.START    

	M.agentColumn =  M.startCol
	M.agentRow = M.startRow

	mine_observation.intArray = [M.startRow * M.col + M.startCol]
	sys.stderr.write('env_start observation %d\n' % (mine_observation.intArray[0]))
	return mine_observation
	
def env_step(action):
	getNextPosition(action) # getNextPosition will update the values of agentRow and agentColumn
 
	mine_observation.intArray = [getPosition()]
  
	mine_ro.o = mine_observation
	mine_ro.r = getReward()
  
	if mine_terminal: # end of episode?
		mine_ro.terminal = 1
	else:
		mine_ro.terminal = 0
	
  	sys.stderr.write('env_step action %d\n' % (action.intArray[0]))
	sys.stderr.write('env_step observation %d\n' % (mine_observation.intArray[0]))
	return mine_ro

def env_cleanup():
	pass

def env_set_state(stateKey):
	pass

def env_set_random_seed(randomSeedKey):
	pass

def env_get_state():
	return State_key()

def env_get_random_seed():
	return Random_seed_key()

def env_message(inMessage):
	return None